File: Model_Selection_Boosting/GridSearch/GridSearch.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Tue Feb 11 10:47:52 2025
@author: swarnabhaghosh
"""

# Grid Search for Hyperparameter Tuning

import numpy as np
import matplotlib.pyplot as plt
import os
import pandas as pd
from sklearn.svm import SVC
from sklearn.model_selection import train_test_split, GridSearchCV, cross_val_score
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import confusion_matrix, accuracy_score
from matplotlib.colors import ListedColormap

# Automatically detect the script directory
script_dir = os.path.dirname(os.path.abspath(__file__))
file_path = os.path.join(script_dir, "Social_Network_Ads.csv")

try:
    dataset = pd.read_csv(file_path) 
    print("Data imported successfully!") 
    X = dataset.iloc[:, [0, 1]].values
    y = dataset.iloc[:, 2].values
except FileNotFoundError:
    print(f"Error: File not found at {file_path}")
    exit()

# Splitting dataset into training & test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=0)

# Feature Scaling
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

# Training Kernel SVM
classifier = SVC(kernel='rbf', random_state=0)
classifier.fit(X_train, y_train)

# Predicting test results
y_pred = classifier.predict(X_test)
print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred))
print("\nAccuracy Score:", accuracy_score(y_test, y_pred))

# Applying K-Fold Cross Validation
accuracies = cross_val_score(estimator=classifier, X=X_train, y=y_train, cv=10)
print("\nK-Fold Cross Validation Mean Accuracy:", accuracies.mean())
print("Standard Deviation:", accuracies.std())

# Applying Grid Search for best parameters
param_grid = [
    {'C': [1, 10, 100, 1000], 'kernel': ['linear']},
    {'C': [1, 10, 100, 1000], 'kernel': ['rbf'], 'gamma': [1, 0.1, 0.01, 0.001]}
]

grid_search = GridSearchCV(estimator=classifier, param_grid=param_grid, scoring='accuracy', cv=10, n_jobs=-1)
grid_search.fit(X_train, y_train)

best_accuracy = grid_search.best_score_
best_params = grid_search.best_params_
print("\nBest GridSearch Accuracy:", best_accuracy)
print("Best Parameters:", best_params)

# Visualizing Training Set Results
X_set, y_set = sc.inverse_transform(X_train), y_train
X1, X2 = np.meshgrid(
    np.arange(start=X_set[:, 0].min() - 10, stop=X_set[:, 0].max() + 10, step=0.25),
    np.arange(start=X_set[:, 1].min() - 1000, stop=X_set[:, 1].max() + 1000, step=0.25)
)

plt.contourf(X1, X2, classifier.predict(sc.transform(np.array([X1.ravel(), X2.ravel()]).T)).reshape(X1.shape),
             alpha=0.75, cmap=ListedColormap(('red', 'green')))
for i, j in enumerate(np.unique(y_set)):
    plt.scatter(X_set[y_set == j, 0], X_set[y_set == j, 1], c=ListedColormap(('red', 'green'))(i), label=j)
plt.title('Kernel SVM (Training Set)')
plt.xlabel('Age')
plt.ylabel('Estimated Salary')
plt.legend()
plt.show()

File: Model_Selection_Boosting/GridSearch/GridSearch.R
# Grid Search with Kernel SVM

# Load Libraries
library(caTools)
library(e1071)
library(caret)
library(ggplot2)

# Import Dataset
dataset <- read.csv('Social_Network_Ads.csv')
dataset <- dataset[, 3:5]

# Encoding Target Variable as Factor
dataset$Purchased <- factor(dataset$Purchased, levels = c(0, 1))

# Splitting Dataset into Training and Test Sets
set.seed(123)
split <- sample.split(dataset$Purchased, SplitRatio = 0.75)
training_set <- subset(dataset, split == TRUE)
test_set <- subset(dataset, split == FALSE)

# Feature Scaling
training_set[, -3] <- scale(training_set[, -3])
test_set[, -3] <- scale(test_set[, -3])

# Training Kernel SVM Model
classifier <- svm(formula = Purchased ~ .,
                  data = training_set,
                  type = 'C-classification',
                  kernel = 'radial')

# Predicting Test Set Results
y_pred <- predict(classifier, newdata = test_set[-3])

# Confusion Matrix
cm <- table(test_set[, 3], y_pred)
print(cm)
accuracy <- sum(diag(cm)) / sum(cm)
print(paste("Test Accuracy:", round(accuracy * 100, 2), "%"))

# Applying K-Fold Cross-Validation
set.seed(123)
folds <- createFolds(training_set$Purchased, k = 10)
cv_results <- lapply(folds, function(x) {
  training_fold <- training_set[-x, ]
  test_fold <- training_set[x, ]
  
  classifier <- svm(formula = Purchased ~ .,
                    data = training_fold,
                    type = 'C-classification',
                    kernel = 'radial')
  
  y_pred_fold <- predict(classifier, newdata = test_fold[-3])
  cm_fold <- table(test_fold[, 3], y_pred_fold)
  accuracy_fold <- sum(diag(cm_fold)) / sum(cm_fold)
  return(accuracy_fold)
})
cv_accuracy <- mean(unlist(cv_results))
print(paste("Cross-Validation Accuracy:", round(cv_accuracy * 100, 2), "%"))

# Applying Grid Search for Hyperparameter Tuning
set.seed(123)
grid_params <- expand.grid(C = c(0.1, 1, 10, 100), sigma = c(0.01, 0.1, 1))

classifier_optimized <- train(Purchased ~ ., 
                              data = training_set, 
                              method = 'svmRadial', 
                              tuneGrid = grid_params, 
                              trControl = trainControl(method = 'cv', number = 10))

print(classifier_optimized$bestTune)

# Function for Visualization
visualize_results <- function(set, title, classifier) {
  X1 <- seq(min(set[, 1]) - 1, max(set[, 1]) + 1, by = 0.01)
  X2 <- seq(min(set[, 2]) - 1, max(set[, 2]) + 1, by = 0.01)
  grid_set <- expand.grid(X1, X2)
  colnames(grid_set) <- c('Age', 'EstimatedSalary')
  
  y_grid <- predict(classifier, newdata = grid_set)
  
  plot(set[, -3],
       main = title,
       xlab = 'Age', ylab = 'Estimated Salary',
       xlim = range(X1), ylim = range(X2))
  contour(X1, X2, matrix(as.numeric(y_grid), length(X1), length(X2)), add = TRUE)
  points(grid_set, pch = '.', col = ifelse(y_grid == 1, 'dodgerblue', 'salmon'))
  points(set, pch = 21, bg = ifelse(set[, 3] == 1, 'dodgerblue3', 'salmon3'))
}

# Visualizing Training Set Results
visualize_results(training_set, "Kernel SVM (Training Set)", classifier_optimized)

# Visualizing Test Set Results
visualize_results(test_set, "Kernel SVM (Test Set)", classifier_optimized)
